edge-hub/src/Microsoft.Azure.Devices.Routing.Core/Dispatcher.cs

```diff
@@ -54,9 +53,11 @@ public static async Task<Dispatcher> CreateAsync(string id, string iotHubName, I
             Preconditions.CheckNotNull(endpointsWithPriorities);
             Preconditions.CheckNotNull(factory);
 
+            var masterCheckpointer = await MasterCheckpointer.CreateAsync(id, NullCheckpointStore.Instance);
+
             IEnumerable<Task<IEndpointExecutor>> tasks = endpointsWithPriorities.Select(e => factory.CreateAsync(e.Key, e.Value));
             IEndpointExecutor[] executors = await Task.WhenAll(tasks);
-            return new Dispatcher(id, iotHubName, executors, factory, new NullCheckpointer());
+            return new Dispatcher(id, iotHubName, executors, factory, masterCheckpointer);
```

Comment:
why do you change from NullCheckpointer to masterCheckpointer? I traced the usages the reason NullCheckpointer is used as isStoreAndForwardEnabled is passed as false in routing module.  So I think it should be kept as NullCheckpointer here, right?

Reply:
it's a mastercheckpointer created on top of a nullcheckpointer store, which has the same effect of not persisting any checkpointing information.  The difference is that NullCheckpointer does not actually do any checkpointing, which broke some of the existing tests that tried to get away with sending the same message multiple-times without changing the offset (as you can see from the test changes).

As far as production goes, there is no valid scenario for disabling StoreAndForward.  That code path is legacy from the original services code and should be considered dead pending removal.

---
In reply to: [390541357](https://github.com/Azure/iotedge/pull/2670#discussion_r390541357) [](ancestors = 390541357)

Reply:
Since that code path is legacy, please create a task/bug to remove it.  thanks.